### PR TITLE
AWS OIDC: List EICE service

### PR DESCRIPTION
**Pull request by marcoandredinis**
This PR creates a new method on the AWS OIDC gRPC service that lists ec2 instance connect endpoints.

This is part of a refactor that moves the API calls behind the Auth Service.

Context: https://github.com/gravitational/teleport/issues/37245

### CODE REVIEW

**Comment by zmb3** [comment]
Out of curiosity - how does this behave when Teleport is not running on AWS? Does the API return an error, is the API not available, etc?

**Comment by marcoandredinis** [comment]
It doesn't use the ambient credentials (like EC2 IAM Profile or any `AWS_` env vars).
It relies entirely on the AWS OIDC Integration, which it uses to generate valid tokens which are then use to issue API calls.

**Comment by zmb3** [comment]
Right, so what happens if I attempt to call this web API or RPC if I haven't configured an AWS OIDC integration? Does it try and fail with an obscure error, or does it tell me that I need to set up an AWS integration first?

**Comment by marcoandredinis** [comment]
It will return `integration not found`

**Review by strideynet — APPROVED** [review]

File: api/proto/teleport/integration/v1/awsoidc_service.proto [hunk 1]
```diff
@@ -32,6 +36,48 @@ service AWSOIDCService {
...
+}
+
+// EC2InstanceConnectEndpoint is a representation of a Amazon VPC EC2 Instance Connect Endpoint.
+message EC2InstanceConnectEndpoint {
+  // Name is the endpoint name.
+  string name = 1;
+  // State is the endpoint state.
+  // Known values:
+  // create-in-progress | create-complete | create-failed | delete-in-progress | delete-complete | delete-failed
+  string state = 2;
```
Comment (Tener) [inline api/proto/teleport/integration/v1/awsoidc_service.proto:59]: Would it make sense to convert this to enum?

File: api/proto/teleport/integration/v1/awsoidc_service.proto [hunk 2]
```diff
@@ -32,6 +36,48 @@ service AWSOIDCService {
...
+message ListEICERequest {
+  // Integration is the AWS OIDC Integration name.
+  string integration = 1;
+  // Region is the AWS Region
+  string region = 2;
+  // VPCIDs is used to filter for EICEs of those VPCs.
+  repeated string vpc_ids = 3;
+  // NextToken is the token to be used to fetch the next page.
+  // If empty, the first page is fetched.
+  string next_token = 4;
```
Comment (Tener) [inline api/proto/teleport/integration/v1/awsoidc_service.proto:49]: Are these all mandatory?

File: api/proto/teleport/integration/v1/awsoidc_service.proto [hunk 3]
```diff
@@ -32,6 +36,48 @@ service AWSOIDCService {
...
+
+// EC2InstanceConnectEndpoint is a representation of a Amazon VPC EC2 Instance Connect Endpoint.
+message EC2InstanceConnectEndpoint {
+  // Name is the endpoint name.
+  string name = 1;
+  // State is the endpoint state.
+  // Known values:
+  // create-in-progress | create-complete | create-failed | delete-in-progress | delete-complete | delete-failed
+  string state = 2;
+  // StateMessage contains a message describing the sate of the EICE.
```
Comment (Tener) [inline api/proto/teleport/integration/v1/awsoidc_service.proto:60]:
```suggestion
  // StateMessage contains a message describing the state of the EICE.
```

File: api/proto/teleport/integration/v1/awsoidc_service.proto [hunk 4]
```diff
@@ -32,6 +36,48 @@ service AWSOIDCService {
...
+  // VPCID is the VPC ID where the Endpoint is created.
+  string vpc_id = 6;
+}
+
+// ListEICEResponse contains a page of AWS EC2 Instance Connect Endpoints.
+message ListEICEResponse {
+  // EC2ICEs contains the page of EC2 Instance Connect Endpoints.
+  repeated EC2InstanceConnectEndpoint ec2ices = 1;
+  // DashboardLink is the URL for AWS Web Console that lists all the Endpoints for the queries VPCs.
+  string dashboard_link = 2;
```
Comment (Tener) [inline api/proto/teleport/integration/v1/awsoidc_service.proto:75]: This field feels weird; how will it change across different pages?

File: api/proto/teleport/integration/v1/awsoidc_service.proto [hunk 5]
```diff
@@ -22,6 +22,10 @@ option go_package = "github.com/gravitational/teleport/api/gen/proto/go/teleport
 
 // AWSOIDCService provides access to AWS APIs using the AWS OIDC Integration.
 service AWSOIDCService {
+  // ListEICE returns a list of EC2 Instance Connect Endpoints.
+  // An optional NextToken that can be used to fetch the next page.
+  rpc ListEICE(ListEICERequest) returns (ListEICEResponse);
```
Comment (Tener) [inline api/proto/teleport/integration/v1/awsoidc_service.proto:27]: nit: Might be worth adding what is the source of this, much like the `ListDatabases` call below lists some APIs.

File: lib/auth/integration/integrationv1/awsoidc.go [hunk 6]
```diff
@@ -200,6 +200,55 @@ func (s *AWSOIDCService) awsClientReq(ctx context.Context, integrationName, regi
 	}, nil
 }
 
+// ListEICE returns a paginated list of EC2 Instance Connect Endpoints.
+func (s *AWSOIDCService) ListEICE(ctx context.Context, req *integrationpb.ListEICERequest) (*integrationpb.ListEICEResponse, error) {
+	authCtx, err := s.authorizer.Authorize(ctx)
```
Comment (Tener) [inline lib/auth/integration/integrationv1/awsoidc.go:205]: I wouldn't mind seeing this tested, if you feel like it.

File: api/proto/teleport/integration/v1/awsoidc_service.proto (hunk 2 above)
Comment (marcoandredinis) [reply api/proto/teleport/integration/v1/awsoidc_service.proto:49]: All except for `next_token`.
I added a comment to `next_token`, expecting that the reader would assume that if I'm being explicit about `next_token` being optional, one would assume the others were required.

I guess it was not effective :sweat_smile: 
Should I add a comment to `integration,region,vpc_ids` fields?

File: api/proto/teleport/integration/v1/awsoidc_service.proto (hunk 1 above)
Comment (Tener) [reply api/proto/teleport/integration/v1/awsoidc_service.proto:74]: Makes sense, I don't think we can easily handle unforeseen values with enum.

File: api/proto/teleport/integration/v1/awsoidc_service.proto (hunk 4 above)
Comment (Tener) [reply api/proto/teleport/integration/v1/awsoidc_service.proto:90]: Actually, now that I look at it we also have the link as part of `EC2InstanceConnectEndpoint`. If this is redundant, we may be able to remove the field from `ListEICEResponse`?

```proto
// DashboardLink is a URL to AWS Console where the user can see the EC2 Instance Connect Endpoint.
  string dashboard_link = 4;
```

**Review by Tener — APPROVED** [review]
Thanks for the changes! I'm approving, but leaving my comment regarding link field for your consideration:
https://github.com/gravitational/teleport/pull/38332/files#r1496011254

File: api/proto/teleport/integration/v1/awsoidc_service.proto (hunk 4 above)
Comment (marcoandredinis) [reply api/proto/teleport/integration/v1/awsoidc_service.proto:90]: > Actually, now that I look at it we also have the link as part of `EC2InstanceConnectEndpoint`. If this is redundant, we may be able to remove the field from `ListEICEResponse`?
> 
> ```proto
> // DashboardLink is a URL to AWS Console where the user can see the EC2 Instance Connect Endpoint.
>   string dashboard_link = 4;
> ```

That one is specific to a single VPC Endpoint, and not for the AWS page where it lists all the EICE endpoints.

Comment (Tener) [reply api/proto/teleport/integration/v1/awsoidc_service.proto:90]: Fair enough!

**Pull request merged**